libbeat/docs/shared-deduplication.asciidoc

```diff
@@ -0,0 +1,139 @@
+[id="{beatname_lc}-deduplication"]
+== Data deduplication
+
+The {beats} framework guarantees at-least-once delivery to ensure that no data
+is lost when events are sent to outputs that support acknowledgement, such as
+{es}, {ls}, Kafka, and Redis. This is great if everything goes as planned. But
+if {beatname_uc} shuts down during processing, or the connection is lost before
+events are acknowledged, you can end up with duplicate data.
+
+[float]
+=== What causes duplicates in {es}?
+
+When an output is blocked, the retry mechanism in {beatname_uc} attempts to
+resend events until they are acknowledged by the output. If the output receives
+the events, but is unable to acknowledge them, the data might be sent to the
+output multiple times. Because document IDs are typically set by {es} _after_ it
+receives the data from {beats}, the duplicate events are indexed as new
+documents.
+
+[float]
+=== How can I avoid duplicates?
+
+Rather than allowing {es} to set the document ID, set the ID in {beats}. The ID
+is stored in the {beats} `@metadata._id` field and used to set the document ID
+during indexing. That way, if {beats} sends the same event to {es} more than
+once, {es} overwrites the existing document rather than creating a new one.
+
+The `@metadata._id` field is passed along with the event so that you can use
+it to set the document ID after the event has been published by {beatname_uc}
+but before it's received by {es}. For example, see <<ls-doc-id>>. 
+
+There are several ways to set the document ID in {beats}:
+
+* *`add_id` processor*
++
+Use the <<add-id,`add_id`>> processor when your data has no natural key field,
+and you can’t derive a unique key from existing fields. 
++
+This example generates a unique ID for each event and adds it to the
+`@metadata._id` field:
++
+[source,yaml]
+----
+processors:
+  - add_id: ~
+----
+ 
+* *`fingerprint` processor*
++
+Use the <<fingerprint,`fingerprint`>> processor to derive a unique key from
+one or more existing fields.
++
+This example uses the values of `field1` and `field2` to derive a unique key
+that it adds to the `@metadata._id` field:
++
+[source,yaml]
+----
+processors:
+  - fingerprint:
+      fields: ["field1", "field2"]
+      target_field: "@metadata._id"
+----
+
+* *`decode_json_field` processor*
```

Comment:
Name of processor should be plural?

Reply:
I get an "F" for attention to detail today. :-P